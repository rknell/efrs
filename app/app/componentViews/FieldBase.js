import React, { Component } from 'react'
import { Text, View } from 'react-native'
import Label from './Label'

export default class FieldBase extends Component {
  render () {
    return (
      <View style={{
        width: this.props.hidden ? 0 : undefined,
        height: this.props.hidden ? 0 : undefined,
        overflow: 'hidden'
      }}>
        <View style={{
          flexDirection: this.props.vertical ? 'column' : 'row',
          borderBottomWidth: this.props.noBorder ? 0 : 1,
          borderColor: this.props.globalStyle ? this.props.globalStyle.neutralColourHighlight : 'rgba(0,0,0,0.1)',
          alignItems: this.props.vertical ? undefined : 'flex-start',
          padding: 10
        }}>
          {!this.props.hideLabel && <Label translate={this.props.translate} left={this.props.vertical} globalStyle={this.props.globalStyle} label={this.props.label} />}
          <View style={{ flex: this.props.vertical ? undefined : 2 }}>{this.props.children}</View>
        </View>
        {!!this.props.error && <Text style={{ margin: 5, color: 'red' }}>{this.props.error}</Text>}
      </View>
    )
  }
}
