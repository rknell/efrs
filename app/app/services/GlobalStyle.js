import { observable } from 'mobx'
import api from '../api'
import { AsyncStorage } from 'react-native'
import tinycolor from 'tinycolor2'

class GlobalStyle {
  style = observable(this._defaultStyle())

  _defaultStyle () {
    return {
      primaryColour: 'dodgerblue',
      neutralColour: 'white',
      secondaryColour: '#f3f3f3',
      fullLogo: require('../assets/efrsFullBanner.png'),
      secondaryLogo: require('../assets/icon.png'),
      errorColour: 'red',
      secondaryColourHighlight: 'black',
      secondaryColourText: 'black',
      secondaryColourTextDisabled: 'black',
      secondaryColourIsDark: 'black',
      secondaryColourDisabled: 'black',
      disabledText: 'black',
      neutralColourHighlight: 'black',
      neutralColourOffset: 'black',
      neutralColourText: 'black',
      neutralColourTextDisabled: 'black',
      neutralColourDisabled: 'black',
      neutralColourIsDark: 'black',
      primaryColourHighlight: 'black',
      primaryColourText: 'black',
      primaryColourTextDisabled: 'black',
      primaryColourDisabled: 'black',
      primaryColourIsDark: 'black'
    }
  }

  constructor () {
    this.loadStyles()
      .then(() => this._recalcStyles())
    this._recalcStyles()
  }

  _lightDark (colour, amount = 15, isDark) {
    const col = tinycolor(colour)
    return (col.isDark() || isDark) ? col.lighten(amount).toRgbString() : col.darken(amount).toRgbString()
  }

  _recalcStyles () {
    this.style.secondaryColourHighlight = this._lightDark(this.style.secondaryColour, undefined, this.style.secondaryIsDark)
    this.style.secondaryColourText = this._lightDark(this.style.secondaryColour, 80, this.style.secondaryIsDark)
    this.style.secondaryColourTextDisabled = tinycolor(this._lightDark(this.style.secondaryColour, 80, this.style.secondaryIsDark)).setAlpha(0.3).toRgbString()
    this.style.secondaryColourIsDark = tinycolor(this.style.secondaryColour).isDark() || this.style.secondaryIsDark
    this.style.secondaryColourDisabled = tinycolor(this.style.secondaryColour).setAlpha(0.3).toRgbString()
    this.style.disabledText = this._lightDark(this.style.neutralColour, 30)
    this.style.neutralColourHighlight = this._lightDark(this.style.neutralColour)
    this.style.neutralColourOffset = this._lightDark(this.style.neutralColour, 5)
    this.style.neutralColourText = this._lightDark(this.style.neutralColour, 80)
    this.style.neutralColourTextDisabled = tinycolor(this._lightDark(this.style.neutralColour, 80)).setAlpha(0.3).toRgbString()
    this.style.neutralColourDisabled = tinycolor(this.style.neutralColour).setAlpha(0.3).toRgbString()
    this.style.neutralColourIsDark = tinycolor(this.style.neutralColour).isDark()
    this.style.primaryColourHighlight = this._lightDark(this.style.primaryColour)
    this.style.primaryColourText = this._lightDark(this.style.primaryColour, 80)
    this.style.primaryColourTextDisabled = tinycolor(this._lightDark(this.style.primaryColour, 80)).setAlpha(0.3).toRgbString()
    this.style.primaryColourDisabled = tinycolor(this.style.primaryColour).setAlpha(0.3).toRgbString()
    this.style.primaryColourIsDark = tinycolor(this.style.primaryColour).isDark()
  }

  loadStyles () {
    return AsyncStorage.getItem(`@EFRS.AppSettings:style`)
      .then(style => {
        if (style) {
          style = JSON.parse(style)
          this.style = observable({ ...this.style, ...style })
          this._recalcStyles()
        }
        return AsyncStorage.getItem(`@login:domain`)
      })
      .then(domain => this.loadRemoteStyle(domain))
      .catch(err => {
        console.error(err)
      })
  }

  loadRemoteStyle (domain) {
    const _this = this
    if (domain) {
      return api.company.getStyle(domain)
        .then(style => {
          if (style) {
            AsyncStorage.setItem(`@EFRS.AppSettings:style`, JSON.stringify(style))
            this.style = observable({ ...this.style, ...style })
            _this._recalcStyles()
            setTimeout(() => _this._recalcStyles(), 100)
          }
        })
    } else {
      return Promise.resolve()
    }
  }

  resetStyles () {
    this.style = observable(this._defaultStyle())
    AsyncStorage.setItem(`@EFRS.AppSettings:style`, JSON.stringify(this.style))
  }
}

export default new GlobalStyle()
