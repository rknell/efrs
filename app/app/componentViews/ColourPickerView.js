import React, { Component } from 'react'
import { ScrollView, Text, TouchableOpacity, View } from 'react-native'
import { observer } from 'mobx-react/native'
import FieldBase from '../componentViews/FieldBase'
import NavBar from '../components/NavBar'
import GlobalState from '../services/GlobalState'
import tinycolor from 'tinycolor2'

@observer
class ColourPickerView extends Component {
  onPress = () => {
    GlobalState.navigation.push({ view: <ColourPickerSelectView {...this.props} /> })
  }

  render () {
    let settings = { Label: this.props.label }
    if (this.props.settings) settings = this.props.settings
    return (
      <FieldBase hidden={this.props.editMode ? false : settings['Hidden']} label={settings.Label} {...this.props}>
        <View style={{ flex: 1, justifyContent: 'center' }}>
          <TouchableOpacity style={{ flex: 1, justifyContent: 'center' }} disabled={this.props.readOnly} onPress={this.onPress}>
            {this.props.value ? <View style={{ backgroundColor: this.props.value, padding: 10 }} />
              : <Text style={{ color: 'dodgerblue' }}>Pick Colour</Text>
            }
          </TouchableOpacity>
        </View>
      </FieldBase>
    )
  }
}

export class ColourPickerSelectView extends Component {
  colours = [
    { name: 'Red', value: '#f34235' },
    { name: 'Pink', value: '#e81d62' },
    { name: 'Purple', value: '#9b26af' },
    { name: 'Indigo', value: '#3e50b4' },
    { name: 'Blue', value: '#2095f2' },
    { name: 'Light Blue', value: '#02a8f3' },
    { name: 'Cyan', value: '#00bbd3' },
    { name: 'Teal', value: '#009587' },
    { name: 'Green', value: '#4bae4f' },
    { name: 'Light Green', value: '#8ac249' },
    { name: 'Lime', value: '#ccdb38' },
    { name: 'Yellow', value: '#feea3a' },
    { name: 'Amber', value: '#fec006' },
    { name: 'Orange', value: '#fe9700' },
    { name: 'Deep Orange', value: '#fe5621' },
    { name: 'Brown', value: '#785447' },
    { name: 'Grey', value: '#9d9d9d' },
    { name: 'Blue Grey', value: '#5f7c8a' },
    { name: 'Black', value: '#000' },
    { name: 'White', value: '#fff' }
  ]

  onPress = (colour) => {
    this.props.onChange(colour.value)
    GlobalState.navigation.pop()
  }

  render () {
    return (
      <View style={{ flex: 1, backgroundColor: 'white' }}>
        <NavBar title='Select Colour' />
        <ScrollView style={{ flex: 1, backgroundColor: 'rgba(0,0,0,0.3)' }}
          contentContainerStyle={{ flexDirection: 'row', flexWrap: 'wrap', justifyContent: 'flex-start' }}>
          {this.colours.map(colour => <ColourBrick colour={colour} key={colour.value} onPress={this.onPress} />)}
        </ScrollView>
      </View>
    )
  }
}

export class ColourBrick extends Component {
  onPress = () => {
    this.props.onPress(this.props.colour)
  }

  render () {
    return (<TouchableOpacity key={this.props.colour.name} onPress={this.onPress}
      style={{ flexShrink: 0, flexGrow: 1, flexBasis: 100, height: 100, flex: 1 }}>
      <View style={{ padding: 10, flex: 1, backgroundColor: this.props.colour.value }}>
        <Text
          style={{ color: tinycolor(this.props.colour.value).isLight() ? 'black' : 'white' }}>{this.props.colour.name}</Text>
      </View>
    </TouchableOpacity>)
  }
}

export default ColourPickerView
