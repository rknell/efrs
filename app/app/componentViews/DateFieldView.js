import React, { Component } from 'react'
import {
  Alert,
  DatePickerAndroid,
  Platform,
  Text,
  TextInput,
  TimePickerAndroid,
  TouchableOpacity,
  View
} from 'react-native'
import FieldBase from './FieldBase'
import IconFa from 'react-native-vector-icons/FontAwesome'
import moment from 'moment'
import { computed, observable } from 'mobx'
import { observer } from 'mobx-react/native'
import Modal from 'react-native-modal'
import IconButton from '../components/IconButton'
import { Calendar } from 'react-native-calendars'
import GlobalStyle from '../services/GlobalStyle'
import Translate from '../services/Translate'

const MONTHS = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec']

@observer
class DateFieldView extends Component {
  @observable model = {
    showPicker: false,
    markedDates: {},
    hour: '',
    minute: '',
    meridiem: '',
    isPickingYear: false,
    isPickingMonth: false,
    year: ''
  }

  @computed get displayLabel () {
    let output
    if (this.props.value) {
      if (this.props.settings['Data Type'] === 'time') {
        output = moment(this.props.value).format('HH:mm')
      } else if (this.props.settings['Data Type'] === 'datetime') {
        output = moment(this.props.value).format('Do MMM YYYY, HH:mm')
      } else {
        output = moment(this.props.value).format('Do MMM YYYY')
      }
    } else {
      if (this.props.settings['Data Type'] === 'time') {
        output = 'Select Time'
      } else if (this.props.settings['Data Type'] === 'datetime') {
        output = 'Select Date & Time'
      } else {
        output = 'Select Date'
      }
    }
    return output
  }

  showPicker () {
    if (!this.props.value) this.props.onChange(new Date())
    if (this.props.settings['Initial Picker'] === 'year') this.model.isPickingYear = true
    if (Platform.OS === 'ios' || this.props.settings['Data Type'] === 'datetime') {
      this.model.showPicker = true
    } else if (Platform.OS === 'android') {
      if (this.props.settings['Data Type'] === 'time') {
        TimePickerAndroid.open({
          hour: moment().hour(),
          minute: moment().minute(),
          is24Hour: true
        })
          .then(result => {
            if (result.action !== TimePickerAndroid.dismissedAction) {
              this.props.onChange(moment(`${result.hour}:${result.minute}}`, 'HH:mm').toDate())
            }
          })
          .catch(err => {
            if (err.message) {
              Alert.alert(err.message)
            } else {
              Alert.alert(Translate.do('Can\'t open time picker'))
            }
          })
      } else { // Its a date picker
        DatePickerAndroid.open({
          date: new Date(this.props.value) || new Date(),
          minDate: this.props.minDate ? new Date(this.props.minDate) : undefined,
          maxDate: this.props.maxDate ? new Date(this.props.maxDate) : undefined
        })
          .then(result => {
            if (result.action !== DatePickerAndroid.dismissedAction) {
              this.props.onChange(moment(`${result.year}-${result.month + 1}-${result.day}`, 'YYYY-MM-DD').toDate())
            }
          }).catch(e => {
            Alert.alert(Translate.do('Error'), e.message || Translate.do('An unknown error occurred'))
          })
      }
    } else {
      Alert.alert(Translate.do('Not Supported'), Translate.do('Current platform does not support the date picker.'))
    }
  }

  getMarkedDates = () => {
    const markedDates = {}
    markedDates[moment(this.props.value).format('YYYY-MM-DD')] = {
      selected: true,
      selectedColor: GlobalStyle.style.primaryColour
    }
    return markedDates
  }

  setDate = (value) => {
    if (this.props.settings['Data Type'] === 'date') {
      this.props.onChange(value.dateString)
    } else {
      if (this.props.value) {
        const mDate = moment(this.props.value).year(value.year).month(value.month - 1).date(value.day)
        this.props.onChange(mDate.toISOString())
      } else {
        this.props.onChange(moment(value.dateString).toISOString())
      }
    }
  }

  setYear = year => {
    this.props.onChange(moment(this.props.value).year(year).toISOString())
    this.model.isPickingYear = false
    this.model.isPickingMonth = true
  }

  setMonth = month => {
    this.props.onChange(moment(this.props.value).month(MONTHS.indexOf(month)).toISOString())
    this.model.isPickingMonth = false
  }

  onChangeHour = (value) => {
    if (value.length === 2) {
      this.refs.minuteInput.focus()
    }
    if (value < 24) {
      this.model.hour = value
    }
    this.onChangeTime()
  }

  onChangeMinute = (value) => {
    if (Number(value) < 60) {
      this.model.minute = value
    }
    this.onChangeTime()
  }

  onChangeTime = () => {
    if (this.model.hour && this.model.minute) {
      let actualHour = this.model.hour
      if (Number(this.model.hour) < 12 && this.model.meridiem === 'pm') {
        actualHour = Number(this.model.hour) + 12
      }
      if (Number(this.model.hour) === 12 && this.model.meridiem === 'am') {
        actualHour = 0
      }

      this.props.onChange(moment(this.props.value).minute(this.model.minute).hour(actualHour).toISOString())
    }
  }

  onHourKeyPress = (evt) => {
    if (evt.nativeEvent.key === 'Enter') {
      this.refs.minuteInput.focus()
    }
  }

  onMinuteKeyPress = (evt) => {
    if (evt.nativeEvent.key === 'Backspace' && this.model.minute.length <= 1) {
      this.refs.hourInput.focus()
    }
  }

  setMeridiem = value => () => {
    this.model.meridiem = value
    this.onChangeTime()
  }

  hidePicker = () => { this.model.showPicker = false }

  constructor (props) {
    super(props)

    if (this.props.value) {
      const momentObj = moment(this.props.value)
      this.model.hour = momentObj.hours().toString()
      this.model.minute = momentObj.minutes().toString()
    }
  }

  render () {
    return (
      <FieldBase hidden={this.props.settings['Hidden'] && !this.props.editMode}
        label={this.props.settings['Label']} {...this.props}>

        {this.props.settings['Set Now Button']
          ? <View style={{ justifyContent: 'center', flex: 1 }}>
            <TouchableOpacity onPress={() => this.props.onChange(new Date())} style={{
              backgroundColor: this.props.globalStyle ? this.props.globalStyle.primaryColour : 'dodgerblue',
              padding: 10,
              alignItems: 'center',
              borderRadius: 10,
              alignSelf: 'stretch'
            }}>
              <Text style={{ color: this.props.globalStyle ? this.props.globalStyle.primaryColourText : 'white' }}>
                {this.props.settings['Data Type'] === 'time' ? Translate.do('Set Time') : Translate.do('Set Date')}
              </Text>
            </TouchableOpacity>
            {!!this.props.value &&
            <Text style={{
              alignSelf: 'center',
              padding: 5,
              color: this.props.globalStyle.neutralColourText
            }}>{this.displayLabel}
            </Text>}
          </View>
          : this.model.showPicker
            ? <Modal
              isVisible={this.model.showPicker}
              swipeDirection='up'
              onSwipe={() => {
                this.model.showPicker = false
              }}>

              <View style={{ margin: 10, padding: 10, backgroundColor: 'white', borderRadius: 10 }}>
                {this.props.settings['Data Type'] !== 'time' &&
                <View style={{ height: 350, justifyContent: 'center' }}>
                  {this.model.isPickingYear &&
                  <TouchableOpacity onPress={() => {
                    this.model.isPickingMonth = true
                    this.model.isPickingYear = false
                  }} style={{ justifyContent: 'center' }}>
                    <View style={{ padding: 30 }}>
                      <Text style={{ fontWeight: 'bold', fontSize: 16 }}>{Translate.do('Select Year')}</Text>
                      <TextInput placeholder='YYYY'
                        value={this.model.year}
                        onChangeText={val => { this.model.year = val }}
                        keyboardType='numeric'
                        maxLength={4}
                        selectTextOnFocus
                        style={{
                          fontSize: 32,
                          textAlign: 'center',
                          borderWidth: 1,
                          borderColor: GlobalStyle.style.neutralColourHighlight
                        }} />
                    </View>
                    <IconButton label='Select Month' name='chevron-right' translate onPress={() => {
                      this.model.isPickingMonth = true
                      this.model.isPickingYear = false
                      this.setYear(this.model.year)
                    }} />
                  </TouchableOpacity>
                  }
                  {this.model.isPickingMonth &&
                  <View style={{ flex: 1, alignItems: 'center', justifyContent: 'center' }}>
                    <Text style={{ fontWeight: 'bold', fontSize: 16 }}>{Translate.do('Select Month')}</Text>
                    <View style={{ flexDirection: 'row', flexWrap: 'wrap' }}>
                      {MONTHS.map(month => <TouchableOpacity
                        style={{ flexBasis: 80, flex: 1, alignItems: 'center', padding: 20, justifyContent: 'center' }}
                        onPress={() => { this.setMonth(month) }}>
                        <Text style={{ color: GlobalStyle.style.secondaryColour }}>{Translate.do(month)}</Text>
                      </TouchableOpacity>)
                      }
                    </View>
                  </View>
                  }
                  {!this.model.isPickingMonth && !this.model.isPickingYear &&
                  <Calendar
                    current={moment(this.props.value).format('YYYY-MM-DD')}
                    minDate={this.props.minDate}
                    maxDate={this.props.maxDate}
                    onDayPress={this.setDate}
                    markedDates={this.getMarkedDates()}
                    onPressYear={() => { this.model.isPickingYear = true; this.model.year = moment(this.props.value).format('YYYY') }}
                    theme={{
                      arrowColor: GlobalStyle.style.secondaryColour
                    }}
                  />
                  }
                </View>
                }

                {(this.props.settings['Data Type'] === 'datetime' || this.props.settings['Data Type'] === 'time') &&
                <View style={{
                  padding: 10,
                  borderTopWidth: 1,
                  borderTopColor: this.props.globalStyle.neutralColourOffset,
                  alignItems: 'center'
                }}>
                  <View>
                    <View style={{ borderBottomColor: this.props.globalStyle.neutralColourOffset, borderBottomWidth: 1 }}>
                      <Text style={{ fontSize: 16 }}>{Translate.do('Set Time')}</Text>
                    </View>
                    <View style={{ flexDirection: 'row', alignItems: 'center' }}>
                      <TextInput ref='hourInput' onChangeText={this.onChangeHour} value={this.model.hour}
                        placeholder='hh'
                        selectTextOnFocus style={{ fontSize: 24, textAlign: 'right', width: 50 }}
                        keyboardType='numeric' maxLength={2}
                        underlineColorAndroid='rgba(0,0,0,0)'
                        onKeyPress={this.onHourKeyPress} /><Text style={{ fontSize: 24 }}> : </Text>
                      <TextInput ref='minuteInput' placeholder='mm' selectTextOnFocus style={{ fontSize: 24 }}
                        underlineColorAndroid='rgba(0,0,0,0)'
                        onChangeText={this.onChangeMinute}
                        value={this.model.minute}
                        keyboardType='numeric' maxLength={2} onKeyPress={this.onMinuteKeyPress} />
                      {Number(this.model.hour) <= 12 &&
                      <View style={{ flexDirection: 'row', marginLeft: 10 }}>
                        <TouchableOpacity onPress={this.setMeridiem('am')}>
                          <Text style={{
                            fontSize: 18,
                            color: this.model.meridiem === 'am' ? this.props.globalStyle.primaryColour : this.props.globalStyle.disabledText
                          }}>am</Text>
                        </TouchableOpacity>
                        <Text
                          style={{ fontSize: 18 }}> / </Text>
                        <TouchableOpacity onPress={this.setMeridiem('pm')}>
                          <Text style={{
                            fontSize: 18,
                            color: this.model.meridiem === 'pm' ? this.props.globalStyle.primaryColour : this.props.globalStyle.disabledText
                          }}>pm</Text>
                        </TouchableOpacity>
                      </View>
                      }

                    </View>
                  </View>

                </View>
                }
                {!this.model.isPickingYear && !this.model.isPickingMonth &&
                <IconButton label='Confirm' name='check' onPress={this.hidePicker} />}
              </View>

            </Modal>
            : <View style={{ flex: 1, alignItems: 'stretch', justifyContent: 'center' }}>
              {!this.props.settings['Read Only']
                ? <TouchableOpacity disabled={this.props.settings['Read Only'] || this.props.readOnly}
                  onPress={() => this.showPicker()} style={{
                    flex: 1,
                    justifyContent: 'center'
                  }}>
                  <Text style={{
                    fontSize: 16,
                    color: this.props.globalStyle ? this.props.globalStyle.primaryColour : '#1E90FF',
                    fontWeight: 'bold'
                  }}>
                    {Translate.do(this.displayLabel)}&nbsp;<IconFa name='calendar' size={20} />
                  </Text>
                </TouchableOpacity>
                : <Text style={{ fontSize: 16, color: 'black', fontWeight: 'bold' }}>{Translate.do(this.displayLabel)}</Text>
              }
            </View>
        }
      </FieldBase>
    )
  }
}

export default DateFieldView
