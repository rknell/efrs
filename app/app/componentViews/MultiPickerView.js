import React, { Component } from 'react'
import { FlatList, StyleSheet, Text, TouchableOpacity, View } from 'react-native'
import FieldBase from './FieldBase'
import GlobalState from '../services/GlobalState'
import IconFa from 'react-native-vector-icons/FontAwesome'
import { observer } from 'mobx-react/native'
import NavBar from '../components/NavBar'
import SearchBar from '../components/SearchBar'
import { observable } from 'mobx'
import GlobalStyle from '../services/GlobalStyle'
import Translate from '../services/Translate'

@observer
class MultiPickerView extends Component {
  viewOptions = () => {
    GlobalState.navigation.push({
      view: <MultiPickerDetail globalStyle={this.props.globalStyle} {...this.props} />
    })
  }

  render () {
    const settings = this.props.settings || { Label: this.props.label }

    return (
      <FieldBase globalStyle={this.props.globalStyle}
        translate={this.props.translate}
        noBorder={this.props.noBorder}
        label={settings.Label}
        hidden={this.props.editMode ? false : settings['Hidden']}>

        <View style={{ flexDirection: 'row' }}>
          <TouchableOpacity onPress={this.viewOptions} disabled={this.props.readOnly}>
            <View style={{ flexDirection: 'row', flexWrap: 'wrap', alignItems: 'flex-start' }}>
              {(!!this.props.values && this.props.values.length && (this.props.values[0] !== ''))
                ? this.props.values.map((item, index) => {
                  return (
                    <View key={index} style={{
                      padding: 10,
                      flex: -1,
                      marginRight: 10,
                      backgroundColor: this.props.globalStyle ? this.props.globalStyle.primaryColour : 'dodgerblue',
                      marginBottom: 10,
                      borderRadius: 8
                    }}><Text style={{
                        color: this.props.globalStyle ? this.props.globalStyle.primaryColourText : 'white',
                        fontWeight: 'bold'
                      }}>{(this.props.translate && !this.props.dontTranslateOptions) ? Translate.do(item.label || item) : item.label || item}</Text></View>)
                })
                : <Text style={{
                  padding: 10,
                  color: this.props.readOnly ? this.props.globalStyle.neutralColourTextDisabled : this.props.globalStyle.neutralColourText
                }}>{Translate.do(`Select an item`)}</Text>
              }
            </View>
          </TouchableOpacity>
        </View>
      </FieldBase>
    )
  }
}

@observer
class MultiPickerDetail extends Component {
  @observable model = {
    searchFilter: ''
  }

  constructor (props) {
    super(props)
    this.onSearchBarChange = this.onSearchBarChange.bind(this)
  }

  // TODO: Find out where the last case is happening (its somewhere in the rules) and fix it
  getOptions () {
    let output
    if (this.props.options) {
      output = this.props.options
    } else if (this.props.extraProps && this.props.extraProps['Options']) {
      output = this.props.extraProps['Options'].values
    } else if (this.props.settings && this.props.settings.Options && this.props.settings.Options.extraProps && this.props.settings.Options.extraProps.values) {
      output = this.props.settings.Options.extraProps.values
    }

    if (this.model.searchFilter) {
      output = output.filter(item => new RegExp(`${this.model.searchFilter}`, 'gi').test(this.getLabel(item)))
    }

    // If it is a form usage and the items are strings, compare if the trimmed values match
    if (this.props.values && this.props.values[0] && this.props.values[0].trim) {
      output.forEach(item => {
        if (item.value && item.value.trim) {
          const trimmedItem = item.value.trim()
          const found = !!this.props.values.find(value => {
            return (value.trim && value.trim() === trimmedItem)
          })
          setImmediate(() => { item.selected = found })
        }
      })
    }

    return output
  }

  onSearchBarChange (value) {
    this.model.searchFilter = value
  }

  _keyExtractor = (item, index) => item.guid || index
  getLabel = (item) => {
    if (item) {
      const label = this.props.labelPath ? item[this.props.labelPath] : item.label
      return (this.props.translate && !this.props.dontTranslateOptions) ? Translate.do(label) : label
    } else {
      return Translate.do('No label')
    }
  }

  render () {
    const options = this.getOptions()

    return (
      <View style={{ flex: 1 }}>
        <NavBar title={this.props.pickerScreenTitle || 'Select Items'} isMenuDisabled />
        <SearchBar onChange={this.onSearchBarChange} value={this.model.searchFilter} />
        <FlatList
          style={{ flex: 1, backgroundColor: this.props.globalStyle.neutralColour }}
          data={options}
          keyExtractor={this._keyExtractor}
          renderItem={({ item, index }) => <MultiPickerItem item={item}
            translate={this.props.translate}
            dontTranslateOptions={this.props.dontTranslateOptions}
            onSelect={() => this.props.onSelect(item, index)}
            label={this.getLabel(item)} />}
        />
      </View>
    )
  }
}

@observer
class MultiPickerItem extends Component {
  _getColour = () => this.props.item.selected ? GlobalStyle.style.primaryColour : GlobalStyle.style.disabledText

  render () {
    return (<TouchableOpacity onPress={this.props.onSelect}>
      <View style={[styles.multiPickerItem, { borderBottomColor: GlobalStyle.style.neutralColourHighlight }]}>
        <Text
          style={{ color: GlobalStyle.style.neutralColourText }}>{(this.props.translate && !this.props.dontTranslateOptions) ? Translate.do(this.props.label) : this.props.label}</Text>
        <View style={{ flex: 1 }} />
        <IconFa name='check' size={20} style={{ color: this._getColour() }} />
      </View>
    </TouchableOpacity>)
  }
}

const styles = StyleSheet.create({
  multiPickerItem: {
    flexDirection: 'row',
    padding: 20,
    borderBottomWidth: 1
  }
})

export default MultiPickerView
export { MultiPickerItem, MultiPickerDetail }
