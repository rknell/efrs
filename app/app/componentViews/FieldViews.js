import DateFieldView from './DateFieldView'
import HeaderFieldView from './HeaderFieldView'
import ListFieldView from './ListFieldView'
import PickerView from './PickerView'
import QuickPickView from './QuickPickView'
import TextFieldView from './TextFieldView'
import ColourPickerView from './ColourPickerView'
import GroupPickerFieldView from './GroupPickerFieldView'
import UserPickerFieldView from './UserPickerFieldView'
import FormFieldPickerView from './FormFieldPickerView'

export default {
  DateFieldView,
  HeaderFieldView,
  ListFieldView,
  PickerView,
  QuickPickView,
  TextFieldView,
  ColourPickerView,
  GroupPickerFieldView,
  UserPickerFieldView,
  FormFieldPickerView
}
