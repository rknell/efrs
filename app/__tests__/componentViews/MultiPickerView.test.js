import React from 'react'
import { MultiPickerDetail, MultiPickerItem } from '../../app/componentViews/MultiPickerView'
import { render, shallow } from 'enzyme'
import GlobalStyle from '../../app/services/GlobalStyle'
import toJSON from 'enzyme-to-json'

it('should render a multi picker item', () => {
  const wrapper = shallow(
    <MultiPickerItem
      item={{ selected: false }}
      label='TEST' />
  )
  expect(toJSON(wrapper)).toMatchSnapshot()
})

it('should render a multi picker item with no label', () => {
  const wrapper = shallow(
    <MultiPickerItem
      item={{ selected: false }}
      label='' />
  )
  expect(toJSON(wrapper)).toMatchSnapshot()
})

it('should render a multi picker item that is selected', () => {
  const wrapper = shallow(
    <MultiPickerItem
      item={{ selected: true }}
      label='TEST' />
  )
  expect(toJSON(wrapper)).toMatchSnapshot()
})

/**
 * OK. So this throws an error that gets fixed when you update react native test renderer.
 * I think it is being thrown by enzyme not supporting alpha versions.
 * Leave skipped until react native is back on non alpha react
 */
it.skip('should render a multi picker detail', () => {
  const wrapper = render(<MultiPickerDetail
    globalStyle={GlobalStyle}
    options={[
      { name: 'Test 1', guid: 1, selected: true },
      { name: 'Test 2', guid: 2 }
    ]}
    labelPath='name'
  />)
  expect(toJSON(wrapper)).toMatchSnapshot()
})
