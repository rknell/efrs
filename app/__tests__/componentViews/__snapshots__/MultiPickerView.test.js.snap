// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render a multi picker detail 1`] = `
<view
  style="flex:1"
>
  <view>
    <view
      style="background-color:dodgerblue;border-bottom-color:rgb(106, 182, 255);border-bottom-width:0;elevation:1px"
    >
      <view
        style="height:22px"
      />
      <view
        style="padding:10px;flex-direction:row"
      >
        <view
          style="opacity:1"
        >
          <view
            style="padding:6px;padding-left:16px;padding-right:16px;align-items:center;justify-content:center"
          >
            <text
              ellipsizemode="tail"
              style="0:[object Object];2:[object Object]"
            >
              
            </text>
          </view>
        </view>
        <text
          ellipsizemode="tail"
          numberoflines="1"
          style="font-weight:300;color:rgb(255, 255, 255);font-size:20px;flex:1"
        >
          Select Items
        </text>
      </view>
    </view>
  </view>
  <rctscrollview
    data="[object Object],[object Object]"
    initialnumtorender="10"
    maxtorenderperbatch="10"
    numcolumns="1"
    scrolleventthrottle="50"
    stickyheaderindices=""
    style="flex:1"
    updatecellsbatchingperiod="50"
    viewabilityconfigcallbackpairs=""
    windowsize="21"
  >
    <view>
      <view>
        <view
          style="opacity:1"
        >
          <view
            style="0:[object Object];1:[object Object]"
          >
            <text
              ellipsizemode="tail"
              style="color:rgb(51, 51, 51)"
            >
              Test 1
            </text>
            <view
              style="flex:1"
            />
            <text
              ellipsizemode="tail"
              style="0:[object Object];1:[object Object];2:[object Object]"
            >
              
            </text>
          </view>
        </view>
      </view>
      <view>
        <view
          style="opacity:1"
        >
          <view
            style="0:[object Object];1:[object Object]"
          >
            <text
              ellipsizemode="tail"
              style="color:rgb(51, 51, 51)"
            >
              Test 2
            </text>
            <view
              style="flex:1"
            />
            <text
              ellipsizemode="tail"
              style="0:[object Object];1:[object Object];2:[object Object]"
            >
              
            </text>
          </view>
        </view>
      </view>
    </view>
  </rctscrollview>
</view>
`;

exports[`should render a multi picker item 1`] = `
<TouchableOpacity
  activeOpacity={0.2}
>
  <View
    style={
      Array [
        Object {
          "borderBottomWidth": 1,
          "flexDirection": "row",
          "padding": 20,
        },
        Object {
          "borderBottomColor": "rgb(217, 217, 217)",
        },
      ]
    }
  >
    <Text
      style={
        Object {
          "color": "rgb(51, 51, 51)",
        }
      }
    >
      TEST
    </Text>
    <View
      style={
        Object {
          "flex": 1,
        }
      }
    />
    <Icon
      allowFontScaling={false}
      name="check"
      size={20}
      style={
        Object {
          "color": "rgb(179, 179, 179)",
        }
      }
    />
  </View>
</TouchableOpacity>
`;

exports[`should render a multi picker item that is selected 1`] = `
<TouchableOpacity
  activeOpacity={0.2}
>
  <View
    style={
      Array [
        Object {
          "borderBottomWidth": 1,
          "flexDirection": "row",
          "padding": 20,
        },
        Object {
          "borderBottomColor": "rgb(217, 217, 217)",
        },
      ]
    }
  >
    <Text
      style={
        Object {
          "color": "rgb(51, 51, 51)",
        }
      }
    >
      TEST
    </Text>
    <View
      style={
        Object {
          "flex": 1,
        }
      }
    />
    <Icon
      allowFontScaling={false}
      name="check"
      size={20}
      style={
        Object {
          "color": "dodgerblue",
        }
      }
    />
  </View>
</TouchableOpacity>
`;

exports[`should render a multi picker item with no label 1`] = `
<TouchableOpacity
  activeOpacity={0.2}
>
  <View
    style={
      Array [
        Object {
          "borderBottomWidth": 1,
          "flexDirection": "row",
          "padding": 20,
        },
        Object {
          "borderBottomColor": "rgb(217, 217, 217)",
        },
      ]
    }
  >
    <Text
      style={
        Object {
          "color": "rgb(51, 51, 51)",
        }
      }
    />
    <View
      style={
        Object {
          "flex": 1,
        }
      }
    />
    <Icon
      allowFontScaling={false}
      name="check"
      size={20}
      style={
        Object {
          "color": "rgb(179, 179, 179)",
        }
      }
    />
  </View>
</TouchableOpacity>
`;
